### PR TITLE
adding tflite implementation with python

### PR DESCRIPTION
**Pull request by wamiqraza**
adding tflite implementation with python

### Summary
🚀📝📸

This pull request adds a new example project of using YOLOv8 with tflite Runtime for object detection. It includes a `main.py` file that implements the core functionality of the project, and a `README.md` file that documents the project and provides usage instructions and examples.

> _`tflite Runtime`_
> _YOLOv8 detects objects_
> _New example code_

### Walkthrough
*  Add a new example project of using YOLOv8 with tflite Runtime ([link](https://github.com/ultralytics/ultralytics/pull/5196/files?diff=unified&w=0#diff-16e3eac8a5def1ab48031e094476c1cea14702072a3f3cf47b39303eebf1dd70R1-R234), [link](https://github.com/ultralytics/ultralytics/pull/5196/files?diff=unified&w=0#diff-df815ef9156f80a4dff3971dad9fd2ecb3576be672e7f7231bfa8c2e420d917aR1-R63))
  * Create the `main.py` file that implements the object detection pipeline using the `YOLOv8` class ([link](https://github.com/ultralytics/ultralytics/pull/5196/files?diff=unified&w=0#diff-16e3eac8a5def1ab48031e094476c1cea14702072a3f3cf47b39303eebf1dd70R1-R234))
  * Write the `README.md` file that explains the project, installation, usage, and output ([link](https://github.com/ultralytics/ultralytics/pull/5196/files?diff=unified&w=0#diff-df815ef9156f80a4dff3971dad9fd2ecb3576be672e7f7231bfa8c2e420d917aR1-R63))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
New example adding YOLOv8 real-time object detection using TensorFlow Lite (tflite) and OpenCV.

### 📊 Key Changes
- 📝 Added a README with installation and usage instructions for YOLOv8 with tflite.
- 🐍 Added `main.py`, a Python script for running YOLOv8 object detection with tflite.
- 🖼️ Included sample output image to demonstrate detection results.
- 🚀 Included a pre-converted YOLOv8 tflite model ready for use.

### 🎯 Purpose & Impact
- 🎉 Provides users with a straightforward way to deploy YOLOv8 object detection in environments where TensorFlow Lite is preferable or necessary.
- 🏃 Enables running object detection on lightweight or resource-constrained devices, potentially expanding accessibility and use cases for YOLOv8.
- 🔍 Encourages the adoption of YOLOv8 by providing additional tools and documentation for its implementation.
- ✔️ May improve the performance of real-time applications by leveraging optimized tflite inference.

### CODE REVIEW

**Comment by wamiqraza** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
@wamiqraza ah, I'm really sorry, I didn't realize this was duplicating existing features. Ultralytics already incorporates automatic TFLite Predict and Val modes. You use TFLite models with Ultralytics the same way you use PyTorch models, and they produce the same exact outputs:

```python
from ultralytics import YOLO

# Load a PyTorch model
model = YOLO('yolov8n.pt')

# Export to TFLite
file = model.export(format='tflite')

# Load the TFLite model
model = YOLO(file)

# Run inference
results = model('path/to/image.jpg')
```

**Comment by wamiqraza** [comment]
> @wamiqraza ah, I'm really sorry, I didn't realize this was duplicating existing features. Ultralytics already incorporates automatic TFLite Predict and Val modes. You use TFLite models with Ultralytics the same way you use PyTorch models, and they produce the same exact outputs:
> 
> ```python
> from ultralytics import YOLO
> 
> # Load a PyTorch model
> model = YOLO('yolov8n.pt')
> 
> # Export to TFLite
> file = model.export(format='tflite')
> 
> # Load the TFLite model
> model = YOLO(file)
> 
> # Run inference
> results = model('path/to/image.jpg')
> ```

Hi @glenn-jocher, Oh, I was not aware if its duplicating the existing features as mentioned but this contribution main idea is that to do the inference on camera by loading model as well along with image inference. For this we will just do only minimal changes specially in the submitted code not sure if we can do live camera inference on the code that you provide above.  

Also I am working on a another version where we can load YOLO tflite model without calling `from ultralytics ... `. Is that will be fine as with regards to the protocol for YOLO or we must use ultralytics dependency. For instance once we quantize the model we need to deploy on some microcontrollers due to limitation we don't need the whole YOLO library to install there so I am working on script that will have the bounding box information and all other relevant information required to do inference.

**Comment by glenn-jocher** [comment]
Hi @wamiqraza, thanks for your reply and for elaborating on your main idea. Running inference on a live camera stream is indeed an interesting feature. Unfortunately, the current setup of this repository is designed to use models created and loaded via Ultralytics, which provides lots of utility and ensures consistency. 

In terms of deploying models on microcontrollers, I understand your concern about the resource limitation and the need for a lightweight solution. While your plans of a separate script that includes the necessary information for inference might work well in a specific low-resource environment, please keep in mind that the goal of this repository is to provide a general solution that covers a range of applications, which necessitates some level of complexity.

We really appreciate your initiative and ideas. It might be beneficial to keep these points in mind when planning future contributions, to ensure that they align well with the orientation of the repository.

**Comment by ddjukic** [comment]
> @wamiqraza ah, I'm really sorry, I didn't realize this was duplicating existing features. Ultralytics already incorporates automatic TFLite Predict and Val modes. You use TFLite models with Ultralytics the same way you use PyTorch models, and they produce the same exact outputs:
> 
> ```python
> from ultralytics import YOLO
> 
> # Load a PyTorch model
> model = YOLO('yolov8n.pt')
> 
> # Export to TFLite
> file = model.export(format='tflite')
> 
> # Load the TFLite model
> model = YOLO(file)
> 
> # Run inference
> results = model('path/to/image.jpg')
> ```

I sadly have to say that this is (once again) not true.
https://github.com/ultralytics/ultralytics/pull/1695
this pull request fixed the 'catastrophic forgetting' and was later merged, however, the problem has been reintroduced.

how to reproduce:
1. train a vanilla yolov8n-seg model for 50 epochs on coco-seg
2. validate
3. export to tflite
4. validate the tflite converted models

the validation metrics of the tflite models are all 0, the model is either not being converted or inferenced correctly.

ultralytics                  8.0.192
torch                        2.0.1
tensorflow-cpu               2.14.0
sng4onnx                     1.0.1
onnx2tf                      1.17.5
onnx-graphsurgeon            0.3.27

one thing i noticed is that the original forgetting fix was tested with onnx2tf 1.9.1 and 1.11.0, however, the ultralytics AutoUpdater now requests onnx2tf>=1.15.4,<=1.17.5.
Not sure if that was ever confirmed as working, and which version of the onnx2tf and ultralytics the support for tflite broke again.

The last version I have the pipeline for training -> converting to int8 quantized tflite working correctly is in an old environment with ultralytics 8.0.74 and onnx2tf 1.9.12.

Any thoughts?

**Comment by glenn-jocher** [comment]
@ddjukic there are no issues with predicting and validating exported TFLite models, we do this every 24 hours and over commit in every PR, so if you see tests passing then TFLite inference works correctly for speed and accuracy. See this PR's Benchmarks CI:
https://github.com/ultralytics/ultralytics/actions/runs/6390168406/job/17342873498?pr=5196

Screenshot 2023-10-04 at 16 43 59

**Comment by wamiqraza** [comment]
@glenn-jocher we can close this issue then.

**Comment by glenn-jocher** [comment]
@wamiqraza hi, thanks for the confirmation. I appreciate your understanding regarding the current scope and capabilities of the project. We will go ahead and close this issue now.
